File: src/com/thoughtworks/ppt/ExtensionMethod.java
    package com.thoughtworks.ppt;

    public class ExtensionMethod {
        public static void main(String[] args) {
            String str = "hello world";
            System.out.println(StringUtil.reverse(str));
            System.out.println(StringUtil.sort(str, false));
            System.out.println(StringUtil.sort(str, true));
        }
    }
    class StringUtil {
        public static String reverse(String input) {
            char[] arr = input.toCharArray();
            for (int i = 0; i < arr.length/2; i++) {
                char temp = arr[i];
                arr[i] = arr[arr.length-1-i];
                arr[arr.length-1-i] = temp;
            }
            return new String(arr);
        }
        public static String sort(String input, boolean desc) {
            char[] arr = input.toCharArray();
            for (int i = 0; i < arr.length - 1; i++) {
                int min = i;
                for (int j = i + 1; j < arr.length; j++) {
                    if (arr[j] < arr[min]) {
                        min = j;
                    }
                }

                char tmp = arr[i];
                arr[i] = arr[min];
                arr[min] = tmp;
            }

            if(desc) {
                return reverse(new String(arr));
            }
            return new String(arr);
        }
    }

File: src/com/thoughtworks/ppt/ExtensionMethod.kt
    package com.thoughtworks.ppt

    fun main() {
        val str: String = "hello world"
        println(str.reverse());
        println(str.sort(false))
        println(str.sort(true))
    }

    fun String.reverse(): String {
        val arr = this.toCharArray()
        for (i in 0 until arr.size / 2) {
            val temp = arr[i]
            arr[i] = arr[arr.size - 1 - i]
            arr[arr.size - 1 - i] = temp
        }
        return String(arr)
    }

    fun String.sort(desc: Boolean): String {
        val arr = this.toCharArray()
        for (i in 0 until arr.size - 1) {
            var min = i
            for (j in i + 1 until arr.size) {
                if (arr[j] < arr[min]) {
                    min = j
                }
            }
            val tmp = arr[i]
            arr[i] = arr[min]
            arr[min] = tmp
        }

        return if (desc) {
            String(arr).reverse()
        } else {
            String(arr)
        }
    }

